cardano-api/test/Test/Cardano/Api/Typed/Value.hs

```diff
@@ -32,14 +33,41 @@ prop_roundtrip_Value_flatten_unflatten =
 -- Remember that Maps cannot have duplicate keys and therefore
 -- we will never go from Value -> ValueNestedRep (via toValueNestedRep) to a
 -- ValueNestedRep with duplicate values.
+
 prop_roundtrip_Value_unflatten_flatten :: Property
 prop_roundtrip_Value_unflatten_flatten =
     property $ do
       v <- forAll genValueNestedRep
-      let v' = valueToNestedRep (valueFromNestedRep v)
-      v `equiv` v'
-  where
-    equiv (ValueNestedRep a) (ValueNestedRep b) = sort a === sort b
+      canonicalise v === valueToNestedRep (valueFromNestedRep v)
+
+canonicalise :: ValueNestedRep -> ValueNestedRep
+canonicalise (ValueNestedRep bundles) =
+  ValueNestedRep . filter removeZeroAda $ mergeDuplicates bundles
+ where
+  mergeDuplicates :: [ValueNestedBundle] -> [ValueNestedBundle]
+  mergeDuplicates l =
+    let allAssets :: [(PolicyId, Map AssetName Quantity)]
+        allAssets = [(pId, qMap) | (ValueNestedBundle pId qMap) <- l]
+
+        allAda :: [Quantity]
+        allAda = [ada | ValueNestedBundleAda ada <- l]
+
+        summedMinted :: Map PolicyId (Map AssetName Quantity)
+        summedMinted = Map.fromListWith (Map.unionWith (<>)) allAssets
+
+        summedAda :: Quantity
+        summedAda = sum allAda
+
+        flattenedMinted = map (uncurry ValueNestedBundle) $ Map.toList summedMinted
+        flattenedAda = ValueNestedBundleAda summedAda
```

Comment:
I think we can probably write this more concisely, but lets solve the sort & generator issues first.